rclpy/rclpy/executors.py

```diff
@@ -0,0 +1,370 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from concurrent.futures import ThreadPoolExecutor as _ThreadPoolExecutor
+import multiprocessing
+from threading import Condition as _Condition
+from threading import Lock as _Lock
+
+from rclpy.callback_groups import ReentrantCallbackGroup as _ReentrantGroup
+from rclpy.constants import S_TO_NS
+from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
+from rclpy.timer import WallTimer as _WallTimer
+from rclpy.utilities import ok
+
+
+class Executor:
+    """
+    A base class for an executor.
+
+    An executor controls the threading model used to process callbacks. Callbacks are units of work
+    like subscription callbacks, timer callbacks, service calls, and received client responses. An
+    executor controls which threads callbacks get executed in.
+
+    A custom executor must define :func:`Executor.spin_once`. If the executor has any cleanup then
+    it should also define :func:`Executor.shutdown`.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._nodes = set()
+        self._nodes_lock = _Lock()
+        # This is triggered when wait_for_ready_callbacks should rebuild the wait list
+        gc, gc_handle = _rclpy.rclpy_create_guard_condition()
+        self._guard_condition = gc
+        self._guard_condition_handle = gc_handle
+        # True if shutdown has been called
+        self._is_shutdown = False
+        # Number of tasks that are being executed
+        self._num_work_executing = 0
+        self._work_condition = _Condition()
+
+    def shutdown(self, timeout_sec=None):
+        """
+        Stop executing callbacks and wait for their completion.
+
+        :param timeout_sec: Seconds to wait. Block forever if None. Don't wait if <= 0
+        :type timeout_sec: float or None
+        :rtype: None
+        """
+        self._is_shutdown = True
+        # Wait for all work to complete
+        if timeout_sec is None or timeout_sec >= 0:
+            with self._work_condition:
+                self._work_condition.wait_for(lambda: self._num_work_executing == 0, timeout_sec)
+        # Clean up stuff that won't be used anymore
+        with self._nodes_lock:
+            self._nodes = set()
+        _rclpy.rclpy_destroy_guard_condition(self._guard_condition)
```

Comment:
Assuming a user wants to call `shutdown` with a timeout of a few seconds. After the function call returns the user doesn't know if the shutdown was actually "successful / clean". So I would imagine the function should return a `bool` to indicate that.

In the case the shutdown wasn't "successful / clean" I wouldn't expect that the nodes and condition are destroyed. What if the user wants to print a message to the console "waited for shutdown, but still ongoing..." and then wait for another amount of time?

Reply:
Fixed in 34c3ed257b2501ce550f3482624bf5f8a94b31e1